lib/modules/manager/poetry/schema.ts

```diff
@@ -35,39 +36,45 @@ const PoetryGitDependency = z
     git: z.string(),
     tag: z.string().optional().catch(undefined),
     version: z.string().optional().catch(undefined),
+    branch: z.string().optional().catch(undefined),
+    rev: z.string().optional().catch(undefined),
   })
-  .transform(({ git, tag, version }): PackageDependency => {
-    if (!tag) {
-      const res: PackageDependency = {
+  .transform(({ git, tag, version, branch, rev }): PackageDependency => {
+    if (tag) {
+      const parsedUrl = parseGitUrl(git);
+      if (parsedUrl.source === 'github.com') {
+        const { owner, name } = parsedUrl;
+        const repo = `${owner}/${name}`;
+        return {
+          datasource: GithubTagsDatasource.id,
+          currentValue: tag,
+          packageName: repo,
+        };
+      } else {
+        return {
+          datasource: GitRefsDatasource.id,
+          currentValue: tag,
+          packageName: git,
+          skipReason: 'git-dependency',
+        };
+      }
+    } else if (rev) {
+      return {
         datasource: GitRefsDatasource.id,
+        currentValue: branch,
+        currentDigest: rev,
+        replaceString: rev,
+        pinDigests: true,
```

Comment:
```suggestion
```

This shouldn't be necessary

Reply:
Without `replaceString`, `yt-dlp` is not updated and I get these logs ☹️. I'll look further into this.

<details>
<summary>Logs</summary>
<pre>
DEBUG: syncBranchState() (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: getBranchPr(renovate/yt-dlp-digest) (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: findPr(renovate/yt-dlp-digest, undefined, open) (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: findPr(renovate/yt-dlp-digest, undefined, closed) (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Found PR #7 (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: branchExists=false (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: dependencyDashboardCheck=undefined (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Check for closed PR because recreating closed PRs is disabled. (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: findPr(renovate/yt-dlp-digest, Update yt-dlp digest to fe6c82c, !open) (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: prAlreadyExisted=false (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Checking schedule(at any time, null) (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: No schedule defined (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Branch needs creating (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Using reuseExistingBranch: false (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Setting current branch to main (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: latest commit (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
       "branchName": "main",
       "latestCommitDate": "2023-12-19T11:49:17-05:00"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
DEBUG: Starting search at index 449 (repository=NextFire/poetry-git-rev-renovate-repro, packageFile=pyproject.toml, branch=renovate/yt-dlp-digest)
       "depName": "yt-dlp"
DEBUG: Found match at index 449 (repository=NextFire/poetry-git-rev-renovate-repro, packageFile=pyproject.toml, branch=renovate/yt-dlp-digest)
       "depName": "yt-dlp"
DEBUG: Digest is not updated (repository=NextFire/poetry-git-rev-renovate-repro, packageFile=pyproject.toml, branch=renovate/yt-dlp-digest)
       "manager": "poetry",
       "expectedValue": "fe6c82ccff6338d97711204162731a8a6741c9b4",
       "foundValue": "2325d03aa7bb80f56ba52cd6992258e44727b424"
 WARN: Error updating branch: update failure (repository=NextFire/poetry-git-rev-renovate-repro, branch=renovate/yt-dlp-digest)
</pre>
</details>

Reply:
So I believe this is due to https://github.com/NextFire/renovate/blob/b69416ce1745f67c9fc1d149738e2f52feb4f732/lib/workers/repository/update/branch/auto-replace.ts#L213.

I see other managers doing the same https://github.com/NextFire/renovate/blob/3c33a2644181cc8e885dd5cb9ce24e23cfab3665/lib/modules/manager/kustomize/extract.ts#L105 so I think `replaceString` is relevant here to force an update of `currentDigest` instead of `currentValue` (which value is required only by the Git Refs datasource).

Reply:
`replaceString` is required for auto replace, otherwise it would try to replace the digest inside the current value.